docs/architecture/ADR-003-Non-interactive-defaults.md

```diff
@@ -0,0 +1,691 @@
+![fit the extra message by removing a tx](./assets/fit-extra-msg-byremoving-tx.png "extra message")
+
+To meet the above constraints, there are multiple refactors required.
+
+- Add metadata to wrapped transactions, connecting that transaction with a message that it pays for.
+- Refactor share splitting and merging to use the above metadata when decoding and encoding the square.
+- Implement the non-interactive default logic.
+- Implement the ability to traverse nmt tree to find subtree roots.
+- Refactor `PrepareProposal` to arrange the shares such that each message has the appropriate subtree roots, and so that the metadata connection between transactions and messages is correct.
+- Refactor `ProcessProposal` to check for message inclusion using only subtree roots to row roots.
+
+### Add metadata to wrapped transactions [#819](https://github.com/celestiaorg/celestia-core/pull/819)
+
+In order to check for message inclusion, create message inclusion fraud proofs, split the block data into squares, and not force non-interactive defaults for every square, we have to connect a `MsgPayForData` transaction to its corresponding message by adding the index of the share that the message starts on as metadata. Since users cannot know this ahead of time, block producers have to add this metadata before the transaction gets included in the block.
+
+We are already wrapping/unwrapping malleated transactions, so including the `share_index` as metadata using the current code is essentially as simple as adding the `share_index` to the struct. Transactions are unwrapped selectively by using a [`MalleatedTxDecoder`](https://github.com/celestiaorg/celestia-app/blob/5ac236fb1dab6628e98a505269f295c18e150b27/app/encoding/malleated_tx_decoder.go#L8-L15) or by using the [`UnwrapMalleatedTx`](https://github.com/celestiaorg/celestia-core/blob/212901fcfc0f5a095683b1836ea9e890cc952dc7/types/tx.go#L214-L237) function.
+
+```proto
+message MalleatedTx {
+bytes original_tx_hash = 1;
+bytes tx               = 2;
+uint32 share_index     = 3;
+}
+```
+
+### Refactor Share Splitting and Merging
+
+Our encoding scheme now has to actually use the metadata added to wrapped transactions.
+
+Note: In order to properly test the new encoding scheme, we have to perform identical if not very similar application logic to that in `PrepareProposal`, for this reason, we initially wanted to move the share encoding/decoding logic to the app instead of core. There were some other quality of life improvements that were also added during this refactor that are technically unrelated to these changes.
+
+We currently utilize a struct to store the state needed for lazily writing message shares. Here we add a method to it that allows for us to write namespaced padded shares.
+
+```go
+// MessageShareSplitter lazily splits messages into shares that will eventually be
+// included in a data square. It also has methods to help progressively count
+// how many shares the messages written take up.
```

Comment:
[Non Blocking][Question]
The count also reflects how many padded shares has been written:  Line 126, no?

Reply:
yeah

Reply:
Does it make sense for the comment to reflect that? 
```suggestion
// how many shares and padding the messages written take up.
```
Or, we expect shares to be messages + paddings?

Reply:
I think it makes more sense without specifying padding.

If we specify padding, it kind of sounds like there are two different numbers, one for shares and one for padding. The implementation treats padding and info shares exactly the same, and I actually think this reflects that in a clearer way.

perhaps in the future we could refactor this to be clearer by only have a single writing method, and then have a function that creates padded shares given a namespace and count. Then the user keeps track of the last namespace.

Reply:
Sounds good, thanks a lot